src/theory/partition_generator.cpp

```diff
@@ -0,0 +1,173 @@
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * PartitionGenerator for creating partitions.
+ */
+
+#include "theory/partition_generator.h"
+
+#include <math.h>
+
+#include "expr/node_algorithm.h"
+#include "expr/node_builder.h"
+#include "options/parallel_options.h"
+#include "prop/prop_engine.h"
+#include "prop/theory_proxy.h"
+#include "prop/zero_level_learner.h"
+#include "theory/theory_engine.h"
+#include "theory/theory_id.h"
+#include "theory/theory_traits.h"
+
+using namespace std;
+using namespace cvc5::theory;
+
+namespace cvc5 {
+
+namespace theory {
+PartitionGenerator::PartitionGenerator(Env& env,
+                                       TheoryEngine* theoryEngine,
+                                       prop::PropEngine* propEngine)
+    : EnvObj(env),
+      d_numPartitions(options().parallel.computePartitions),
+      d_numChecks(0),
+      d_numPartitionsSoFar(0)
+{
+  d_valuation = std::make_unique<Valuation>(theoryEngine);
+  d_propEngine = propEngine;
+
+  d_conflictSize = options().parallel.partitionConflictSize;
+  if (!d_conflictSize)
+  {
+    d_conflictSize = (uint64_t)log2(d_numPartitions);
+  }
+}
+
+void PartitionGenerator::collectDecisionLiterals(std::vector<TNode>& literals)
+{
+  std::vector<Node> decisionNodes = d_propEngine->getPropDecisions();
+  // Make sure the literal does not have a boolean term or skolem in it.
+  const std::unordered_set<Kind, kind::KindHashFunction> kinds = {
+      kind::SKOLEM, kind::BOOLEAN_TERM_VARIABLE, kind::CONST_BOOLEAN};
+
+  for (const Node& n : decisionNodes)
+  {
+    Node originalN = SkolemManager::getOriginalForm(n);
+
+    // If the literal is the not of some node, do the checks for the child
+    // of the not instead of the not itself.
+    Node original = originalN.getKind() == kind::NOT ? originalN[0] : originalN;
+    if (expr::hasSubtermKinds(kinds, original)
+        || !d_valuation->isSatLiteral(original)
+        || !d_valuation->isDecision(original)
+        || Theory::theoryOf(original) == THEORY_BOOL)
+    {
+      continue;
+    }
+
+    literals.push_back(originalN);
+  }
+}
+
+void PartitionGenerator::emitCube(Node toEmit)
+{
+  *options().parallel.partitionsOut << toEmit << std::endl;
+  ++d_numPartitionsSoFar;
+}
+
+TrustNode PartitionGenerator::blockPath(TNode toBlock)
+{
+  // Now block the path in the search.
+  Node lemma = toBlock.notNode();
+  d_assertedLemmas.push_back(lemma);
+  TrustNode trustedLemma = TrustNode::mkTrustLemma(lemma);
+  return trustedLemma;
+}
+
+// Send lemma that is the negation of all previously asserted lemmas.
+TrustNode PartitionGenerator::stopPartitioning()
+{
+  std::vector<Node> unsat = {NodeManager::currentNM()->mkConst(true),
+                             NodeManager::currentNM()->mkConst(false)};
```

Comment:
We generally avoid writing `NodeManager::currentNM()` multiple times and just assign it to `nm` somewhere in the scope.